internal/xds/rbac/rbac_engine.go

```
@@ -238,3 +274,30 @@ type rpcData struct {
 	// handshake.
 	certs []*x509.Certificate
 }
+
+func doAuditLogging(engine *engine, rpcData *rpcData, rule string, authorized bool) error {
+	event := &audit.Event{
+		FullMethodName: rpcData.fullMethod,
+		Principal:      rpcData.peerInfo.Addr.String(),
+		PolicyName:     engine.policyName,
+		MatchedRule:    rule,
+		Authorized:     authorized,
+	}
+	for _, logger := range engine.auditLoggers {
+		switch engine.auditCondition {
+		case v3rbacpb.RBAC_AuditLoggingOptions_NONE:
+			continue
+		case v3rbacpb.RBAC_AuditLoggingOptions_ON_DENY:
+			if !authorized {
+				(*logger).Log(event)
+			}
+		case v3rbacpb.RBAC_AuditLoggingOptions_ON_ALLOW:
+			if authorized {
+				(*logger).Log(event)
+			}
+		case v3rbacpb.RBAC_AuditLoggingOptions_ON_DENY_AND_ALLOW:
+			(*logger).Log(event)
+		}
+	}
+	return nil
```

Why does it need to return an error?

---

Good catch, it doesn't